### DISCUSSION TITLE
Create a `<Drawer>` Layout with nested `<Stack>`

### DISCUSSION THREAD

**Discussion by Emperor-Koala** [expo/router#715] [Q&A]
I'm trying to make a `<Drawer>` layout, where one of the routes has a nested `<Stack>`. The main issue I'm encountering is a sort of "conflict" with the headers. I want the root page of the `<Stack>` to include the "open drawer" button in the header, but for any sub-pages of the Stack to just have a back button.

Unfortunately right now if I want to be able to go back up through the stack, I have to show the headers for both navigators, which obviously doesn't look that great.

I tried defining the options on both `<Drawer.Screen />` and `<Stack.Screen />` at the same time, but it just merged them and applied all the options to the Stack navigator. I also thought about using `navigation.openDrawer()`, but typescript complains because `openDrawer()` doesn't exist on the base `useNavigation()`.

Has anyone managed to figure out how to accomplish something like this? Any help would be greatly appreciated!

**Comment by wcastand** [comment]
hope this helps
https://github.com/wcastand/example-drawer

**Reply by Emperor-Koala** [reply]
Unfortunately, I ended up starting on my current project about a month and a half too early. I'm using Expo SDK 48 and Expo Router v1.

**Reply by NaumanJafarSintov** [reply]
@Emperor-Koala did you find a solution to this problem, I am facing the same problem

**Reply by NaumanJafarSintov** [reply]
@Emperor-Koala this issue has the simplest solution, just make the headerShown false for your specific screen inside your drawer and use DrawerToggleButton from react-native/drawer

import { DrawerToggleButton } from "@react-navigation/drawer";
import { Stack } from "expo-router";

export default function StackLayout(){
    return <Stack>
        <Stack.Screen name = "packagesScreen" options={{
            headerLeft: () => <DrawerToggleButton/>
        }}/>
        <Stack.Screen name = "deliveryDetailsScreen"/>
    </Stack>
}

**Comment by sajjadjaved01** [comment]
Hope this helps.

```
import { DrawerActions } from "@react-navigation/native";
import { useNavigation, useRouter } from "expo-router";

  const nav = useNavigation();
  
  <Icon
          onPress={() => {
            nav.dispatch(DrawerActions.toggleDrawer());
          }}
          name="menu"
          size={RFValue(25)}
          color="#040415"
        />
```

**Comment by sandeep194920** [comment]
**Here's how you can create a Drawer + Stack Navigator using Expo router**

Overview - Drawer navigation would be the base navigation and we add Stack Navigation on top of that for this to work.

Step 1. Create an `app/_layout.tsx` file with Drawer Navigation.

```tsx
import React from "react";
import Drawer from "expo-router/drawer";
import { GestureHandlerRootView } from "react-native-gesture-handler";

const RootLayout = () => {
  return (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <Drawer screenOptions={{ headerShown: false }}>
        <Drawer.Screen name="(stacks)" />
      </Drawer>
    </GestureHandlerRootView>
  );
};

export default RootLayout;
```
**NOTE:** Do not create an `index` file inside `/app`. This file is not required because we directly want to render stack layout (shown in step 2) on to the main layout.

Step 2.  Create a folder where you want your stack navigation. For example, `app/(stacks)`. Treat this as your main navigation folder for the app.

File - `app/(stacks)/_layout.tsx`

```tsx
import React from "react";
import { Stack, useNavigation } from "expo-router";
import { Ionicons } from "@expo/vector-icons";
import { DrawerActions } from "@react-navigation/native";

const StacksLayout = () => {
  const nav = useNavigation();
  return (
    <Stack>
      <Stack.Screen
        name="index"
        options={{
          headerTitle: "Stack Home",
          headerLeft: () => {
            return (
              <Ionicons
                name="menu"
                size={24}
                onPress={() => {
                  nav.dispatch(DrawerActions.openDrawer());
                }}
              />
            );
          },
        }}
      />
    </Stack>
  );
};

export default StacksLayout;

```

`NOTE`: Here we add functionality to open drawer and hide the main drawer screen. 

Step 3. Add an index file. `app/(stacks)/index.tsx`

```tsx
import { View, Text } from "react-native";
import React from "react";
import { useRouter } from "expo-router";

const StackHome = () => {
  const router = useRouter();
  return (
    <View>
      <Text>Stack Home</Text>
      <Text
        onPress={() => {
          router.push("/(stacks)/users");
        }}
      >
        Go to users page
      </Text>
    </View>
  );
};

export default StackHome;
```

Step 4. Create a users file inside (stacks). File - `app/(stacks)/users.tsx`

```tsx
import { View, Text } from "react-native";
import React from "react";

const StackUsers = () => {
  return (
    <View>
      <Text>Users Screen</Text>
    </View>
  );
};

export default StackUsers;
```


See the demo below.




https://github.com/user-attachments/assets/a383ffbb-1a29-4c0a-83d9-b4df9e967eb1

**Reply by fabiohbarbosa** [reply]
Ok, that looks nice and it's super helpful

I have got a question:
How can I add or edit the set of routes into the drawer? I can see only the option "(stacks)" on your drawer

**Reply by sandeep194920** [reply]
Hey @fabiohbarbosa, thanks for the reply.  2 things that can be done here as follows to modify the route names in the drawer:

**1. If you want the route name to be the screen name (or any other name).**

You can add `options` inside `Drawer.Screen` and set `drawerLabel` like this  

```jsx
<Drawer.Screen name="(stacks)" options={{ drawerLabel: "Home" }} />
```
This would change '(stacks)' to 'Home'


**2. If you want customized drawer to contain more content  like custom route names, sign-out button and so on(which I believe you would need at some point)**

This is a better approach to customize your drawer. You need to create a custom component for you drawer and then assign it to a prop called `drawerContent` inside `<Drawer  ..here...` like this. So the `Step 1` can be modified as follows:

Step 1. Modify `app/_layout.tsx` file with Drawer Navigation to add `drawerContent` prop.

```jsx
import React from "react";
import Drawer from "expo-router/drawer";
import { GestureHandlerRootView } from "react-native-gesture-handler";
import DrawerContent from "@/components/DrawerContent";

const RootLayout = () => {
  return (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <Drawer
        drawerContent={DrawerContent}
        screenOptions={{ headerShown: false }}
      >
        <Drawer.Screen name="(stacks)" />
      </Drawer>
    </GestureHandlerRootView>
  );
};

export default RootLayout;
```


Step 2. Create a file `components/DrawerContent.tsx` and import into above step.


```jsx
import { StyleSheet, Text } from "react-native";
import React from "react";
import { DrawerContentScrollView, DrawerItem } from "@react-navigation/drawer";
import { router, useNavigation } from "expo-router";
import { DrawerActions } from "@react-navigation/native";
import Icon from "react-native-vector-icons/MaterialIcons";

const DrawerContent = () => {
  const nav = useNavigation();
  return (
    <DrawerContentScrollView style={styles.container}>
      <Text>This is a custom drawer</Text>
      <DrawerItem
        label="Home"
        onPress={() => {
          nav.dispatch(DrawerActions.closeDrawer);
        }}
      />

      <DrawerItem
        label="Users"
        onPress={() => {
          router.push("/(app)/users");
        }}
      />

      <DrawerItem
        label="Sign Out"
        icon={({ color, size }) => (
          <Icon name="person" color="black" size={size} />
        )}
        onPress={() => {}} // logout user function call
        inactiveTintColor="blue"
      />
    </DrawerContentScrollView>
  );
};

export default DrawerContent;

const styles = StyleSheet.create({
  container: {
    backgroundColor: "#D3D3D3",
  },
});

```

An additional tip if you want to modify the screen name within the component. For example, if you want `users` screen to be named `My Users`, you can do it like this inside `users` route.

This is helpful when you want to set the route name dynamically based on a prop/context value.

```jsx
import { View, Text } from "react-native";
import React from "react";
import { Stack } from "expo-router";

const StackUsers = () => {
  return (
    <>
      <Stack.Screen options={{ headerTitle: "My Users" }} />
      <View>
        <Text>Users Screen</Text>
      </View>
    </>
  );
};

export default StackUsers;
```

*NOTE: Just to setup this drawer as per expo router, we might need expo docs, but all the customizations related to drawer can still be found in [reactnavigation docs](https://reactnavigation.org/docs/drawer-navigator#drawercontent)*

https://github.com/user-attachments/assets/1a1266bc-bf22-4ac4-abf4-776cb3712b04


*You can find [full code](https://github.com/sandeep194920/react-native-starters/tree/branch-%231A-EXPO-ROUTER-starter-with-stack-drawer-navigation-setup) for this here. [In this repo](https://github.com/sandeep194920/react-native-starters), each branch contains quick starter files for setting up react-native or expo.*

**Reply by Eresnen** [reply]
Hello, is it possible to update your code with SDK 52 please? because on SDK 52 your code does not work correctly. THANKS

**Reply by sandeep194920** [reply]
@Eresnen what's the issue you are facing with current version?

**Reply by Eresnen** [reply]
With SDK 52, it is not possible to use router.push to go to the users page when you are in index.tsx of the (app) folder. No route changes are made and no errors are displayed

**Reply by sandeep194920** [reply]
Please fork my repo and try if it works for you in my current version. I checked at the [expo documentation](https://docs.expo.dev/versions/latest/sdk/router/#userouter) and seems like `router.push` remains the same in the latest version as well.

**Reply by Eresnen** [reply]
In your version it is functional but not when we upgrade to SDK 52. From now on it is necessary to use SDK 52 and I would have liked to use your repo in version 52.
I mean if we use a Tabs, we can't navigate between tabs using router.push from the drawer

**Reply by sandeep194920** [reply]
Will take a look and get back to you.

**Reply by Eresnen** [reply]
Thank's

**Reply by carsherman-paypal** [reply]
bump
